project_group_setup.py

```diff
@@ -0,0 +1,270 @@
+    try:
+        # All project osggids
+        project_osggids = set(
+            int(utils.identifier_from_list(project["ID_List"], "osggid")) for project in project_groups
+        )
+        # project osggids not provisioned in ldap
+        project_osggids_to_provision = project_osggids.difference(ldap_group_osggids)
+        # All projects that aren't provisioned in ldap
+        projects_to_provision = (
+            project
+            for project in project_groups
+            if int(utils.identifier_from_list(project["ID_List"], "osggid")) in project_osggids_to_provision
+        )
+        return projects_to_provision
+    except TypeError:
+        return set()
+
+
+def add_missing_group_identifier(project, id_type, value):
+    # If the group doesn't already have an id of this type ...
+    if utils.identifier_from_list(project["ID_List"], id_type) is None:
+        # ... add the identifier to the group
+        utils.add_identifier_to_group(project["Gid"], id_type, value, options.endpoint, options.authstr)
+        print(f'project {project["Gid"]}: added id {value} of type {id_type}')
+
+
+def assign_identifiers_to_project(project, id_dict):
+    for k, v in id_dict.items():
+        # Add an identifier of type k and value v to this group, if it doesn't have them already
+        add_missing_group_identifier(project, k, v)
+    # Update the project object to include the new identifiers
+    new_identifiers = utils.get_co_group_identifiers(project["Gid"], options.endpoint, options.authstr)["Identifiers"]
+    project["ID_List"] = new_identifiers
+
+
+def assign_identifiers(project_list, highest_osggid):
+    highest = highest_osggid
+    for project in project_list:
+        # Project name identifier is the CO Group name in lower case
+        project_name = project["Name"].lower()
+
+        # Determine what osggid to assign this project,
+        # based on the starting range and the highest osggid seen in existing groups
+        osggid_to_assign = max(highest + 1, options.project_gid_startval)
+        highest = osggid_to_assign
+
+        identifiers_to_add = {"osggid": osggid_to_assign, "osggroup": project_name}
+
+        assign_identifiers_to_project(project, identifiers_to_add)
+
+
+def create_unix_cluster_groups(project_list):
+    for project in project_list:
+        utils.add_unix_cluster_group(project["Gid"], options.ucid, options.endpoint, options.authstr)
+        print(f'project group {project["Gid"]}: added UNIX Cluster Group')
+
+
+def provision_groups(project_list):
+    for project in project_list:
+        utils.provision_group(project["Gid"], options.provision_target, options.endpoint, options.authstr)
+        print(f'project group {project["Gid"]}: Provisioned Group')
+
+
+def main(args):
+    parse_options(args)
+
+    # Make all of the nessisary calls to COManage's API for the data we'll need to set up projects.
```

Comment:
```suggestion
    # Make all of the necessary calls to COManage's API for the data we'll need to set up projects.
```

I find a spellcheck plugin to be very helpful for comments!